### ISSUE TITLE
Filter can't be applied to type: MultiPolygon

### ISSUE THREAD

**Issue by dreadedhamish** [maplibre/maplibre-gl-js#3516]
Perhaps this is a feature request, but it seems like an existing function is incomplete.

The scenario is "Add multiple geometries from one GeoJSON source" (https://maplibre.org/maplibre-gl-js/docs/examples/multiple-geometries/) but where a Feature is a MultiPolygon.

When applying styling using a filter to a Feature that is of type = MultiPolygon:

`
map.addLayer({
            'id': 'park-boundary',
            'type': 'fill',
            'source': 'national-park',
            'paint': {
                'fill-color': '#888888',
                'fill-opacity': 0.8
            },
            'filter': ['==', '$type', 'MultiPolygon']
        });
`

an error is thrown:

`Error: layers.park-boundary.filter[2]: expected one of [Point, LineString, Polygon], "MultiPolygon" found`

MultiPolygons are styled correctly when not using the filter.

**Comment by HarelM** [comment]
Have you tried `geometry-type`?
https://maplibre.org/maplibre-style-spec/expressions/#geometry-type
When in doubt, I recommend going through the above site, it is very elaborate.

**Comment by dreadedhamish** [comment]
> Have you tried `geometry-type`? https://maplibre.org/maplibre-style-spec/expressions/#geometry-type When in doubt, I recommend going through the above site, it is very elaborate.

Thanks for your help.

Could you help me with expression? I'm a little out of my depth.
I tried:
'filter': ['==', '$type', 'MultiPolygon']
'filter': ["==", ["geometry-type"], "MultiPolygon"] 
'filter': ["match", ["geometry-type"], "MultiPolygon", true, false]
None throw and error, and none work.

The geoJSON is loaded like the example in my original post:
`
map.on('load', () => {
        map.addSource('national-park', {
            'type': 'geojson',
            'data': {
                'type': 'FeatureCollection',
                'features': [
                    {
                        'type': 'Feature',
                        'geometry': {"type": "MultiPolygon", "coordinates":
`

**Comment by HarelM** [comment]
Hmm...
Seems like `geometry-type` is also returning a Polygon instead of a MultiPlygon... :-/
https://jsbin.com/cuhikisela/edit?html,output

This looks like a bug as it looks like geojson-vt support multipplygon:
https://github.com/mapbox/geojson-vt/issues/5

The following is the only test that checks `geometry-type`, which is very limited.
https://github.com/maplibre/maplibre-gl-js/blob/39003c93fcef8b6cf06520e0de391042ca1afbb1/test/integration/expression/tests/geometry-type/basic/test.json#L4

Feel free to try and tackle this, I hope it won't be complicated...

**Comment by wipfli** [comment]
Is that related to this: https://osmus.slack.com/archives/C031V9E9RMG/p1704407869123399?thread_ts=1704405329.375309&cid=C031V9E9RMG ?

**Comment by HarelM** [comment]
It is related, but it seems like a reverse behavior to what's reported here.
Same "area" of the bug but possible a different bug.

**Comment by acalcutt** [comment]
Do you need to distinguish between MultiPolygon and Polygon? In the past when I was testing with buildings, the filter had to be "Polygon" even though the object was a MultiPolygon $type when i viewed in inspect. I just assumed it was matching without the "Multi", since it does't consider those valid geometry types (like shown here in [slack](https://osmus.slack.com/archives/C031V9E9RMG/p1704409539753589?thread_ts=1704405329.375309&cid=C031V9E9RMG) https://github.com/maplibre/maplibre-style-spec/blob/main/src/expression/evaluation_context.ts#L6 )

**Comment by zstadler** [comment]
The style spec contains two means to access the geometry type of an element, each has a different set of possible velues:

1. [`["geometry-type"]`](https://maplibre.org/maplibre-style-spec/expressions/#geometry-type) Gets the feature's geometry type: `Point`, `MultiPoint`, `LineString`, `MultiLineString`, `Polygon`, `MultiPolygon`.
2. [`"$type"`](https://maplibre.org/maplibre-style-spec/deprecations/#combining-filters): the feature type. This key may be used with the `"=="`,`"!="`, `"in"`, and `"!in"` operators. Possible values are `"Point"`, `"LineString"`, and `"Polygon"`.

The values returned by `"$type"` are closely related to the basic [encoding of geometries in the MVT format](https://github.com/mapbox/vector-tile-spec/blob/master/2.1/README.md#434-geometry-types).

On the other hand, the MVT format spec [also says](https://github.com/mapbox/vector-tile-spec/blob/master/2.1/README.md#4342-point-geometry-type) that _a geometry MUST interpreted as a multi* geometry when the number of `MoveTo` commands is greater that 1_. Apparently, this is not part of the the `["geometry-type"]` implementation and therefore it never returns any of the `Multi*` values.

Note that once the `["geometry-type"]` implementation does return `Multi*` values, the implementation of the `"=="`,`"!="`, `"in"`, and `"!in"` operators that use `$type` should also be updated accordingly. For example, `["==", $type, "LineString"]` should be implemented as `["in", ["geometry-type"], "LineString", "MultiLineString"]` rather than `["==", ["geometry-type"], "LineString"]` as currently done by `gl-style-migrate`.

**Comment by HarelM** [comment]
What's more interesting for me, is to understand if this happens to MPs in MVT.
If this is only an issue with geojson source, then the fix should be in the geojson-vt library and not here.
But if there's a wider issue that also applies to mvt, then it needs to be solved.

**Comment by zstadler** [comment]
The problem, as far as I understand, is unrelated to GeoJSON. Here is an example where the vector tiles were created from OSM data by OpenMapTiles.

Here is an inner member of an OSM nature reserve boundary relation where `["element-type"] == "Polygon"` is shown on the member's geometry, rather than `"MultiPolygon "`

![image](https://github.com/maplibre/maplibre-gl-js/assets/1304610/1a07fe36-757d-4dfc-8dc6-09cf70247d90)

**Comment by sbachinin** [comment]
@HarelM, apparently it's a wider issue that also applies to mvt.

Type "MP" gets lost with VT sources too. It gets lost not within Maplibre but supposedly when tiles are generated.

I tried a couple of ways to generate tiles (tippecanoe from geojson, maptiler from osm data) and in both cases MP was stored as P, i.e. the "type" of a feature was renamed from "MP" to "P".
("The POLYGON geometry type encodes a polygon or multipolygon geometry", mvt spec says).

Though the structure of an original MP is preserved. The resulting "P" (ex-MP) feature still contains multiple "lines" that correspond to original constituent Ps. (But not necessarily multiple. Depends on tiles' structure).

On the frontend nothing intersting happens. @mapbox/vector-tile basically just decodes the fetched PBF and makes no assumptions about feature types. I.e., it just takes the hardcoded feature type ("3") from the appropriate index in PBF.

I hope this answers your question. (https://github.com/maplibre/maplibre-gl-js/issues/3516#issuecomment-1892341938).

Does it mean that this problem should be solved in maplibre?
I think it might be possible (though very complicated) with geoJSON sources.
And I doubt that it's possible for VT sources because it seems that VTs doesn't give enough information to deduce the original feature type.
Anyway to build a real solution would mean working against the overall framework of mvt. Or perhaps there is no framework really, but MPs are just under-supported.

**Comment by zstadler** [comment]
@sbachinin, @HarelM,

The MVT Spec [also says](https://github.com/mapbox/vector-tile-spec/blob/master/2.1/README.md#4344-polygon-geometry-type) (my emphasis on **MUST**):
> If the command sequence for a POLYGON geometry type includes only a single exterior ring then the geometry **MUST** be interpreted as a single polygon; otherwise the geometry **MUST** be interpreted as a multipolygon geometry, wherein each exterior ring signals the beginning of a new polygon. If a polygon has interior rings they MUST be encoded directly after the exterior ring of the polygon to which they belong.

It looks like the interpretation required by the MVT Spec, using a count of external rings, is not performed at any point during the processing of the vector tiles. The numeric values provided are [converted to their literal string names](https://github.com/maplibre/maplibre-style-spec/blob/61dff6a58ff4087e51fb02d02b3a006dbc2f00c9/src/expression/evaluation_context.ts#L32-L34).

On the other  hand, [VectorTileFeature.prototype.toGeoJSON()](https://github.com/mapbox/vector-tile-js/blob/77851380b63b07fd0af3d5a3f144cc86fb39fdd1/lib/vectortilefeature.js#L129) performs that check by [classifying the rings](https://github.com/mapbox/vector-tile-js/blob/77851380b63b07fd0af3d5a3f144cc86fb39fdd1/lib/vectortilefeature.js#L197) and [counting the number of outer rings](https://github.com/mapbox/vector-tile-js/blob/77851380b63b07fd0af3d5a3f144cc86fb39fdd1/lib/vectortilefeature.js#L173-L177). Similar treatment is done for MultiPoint and MultiLineString features.

**Comment by zstadler** [comment]
Here is a proposal to sort this out:

- `["geometry-type"]` will always give the detailed geometry, including the classification of MVT features as "Multi"* geometries.
- `"$type"` will always give the simplified geometry by removing any "Multi" prefix of a geometry type.

With this proposal, 
- The two type operators of the style spec will provide their expected values regardless of the feature source - MVT or GeoJSON
- No changes are required to the style spec
- maplibre-gl-js will comply with the interpretation required by MVT Spec

Implementation thoughts
- The evaluation of  `["geometry-type"]` can be lazy in the sense that the classification of MVT features will be done only once, and the string result of the classification will replace the numeric `.type`. That will avoid the need for subsequent classifications of the same feature.
- The evaluation of `"$type"` will continue to handle numeric `.type` values and will not trigger their classification. It will only remove any `"Multi"` prefix from string `.type` values.

**Comment by HarelM** [comment]
The above proposal seems right to me as it aligns the implantation to the spec.
@louwers can you check what is happening in native for the two filters discussed here?

**Comment by zstadler** [comment]
I think we have been discussing this issue under an assumption that `["geometry-type"]` differentiates between `Polygon` and `MultiPolygon`. Now, I'm not sure that `["geometry-type"]` will ever produce a `"MultiPolygon"` value.

@dreadedhamish,
Could you please add the following layer as the last layer in a style you're using and see what `["geometry-type"]` value are placed on the GeoJGON's MultiPolygons of the `national-park` layer?

```
{
  "id": "geometry-type",
  "type": "symbol",
  "source": "national-park",
  "layout": {
    "text-field": ["geometry-type"],
    "text-font": ["Open Sans Regular"]
  }
}
```

So far, I was unable to produce an example that shows a `MultiPolygon` value of `["geometry-type"]`, even when inlining GeoJSON in the style itself.
https://codepen.io/zstadler/pen/LYajQPx?editors=0010

![image](https://github.com/maplibre/maplibre-gl-js/assets/1304610/414d5de2-d6f2-439e-8645-a13d6bbd4d7d)

**Comment by zstadler** [comment]
I've modified the "Maptiler Basic" style as described above.

When opening [this style](https://github.com/maplibre/maplibre-gl-js/files/14125490/basic.json) in Maplibre Maputnik, the Inspect view says the `$type` of the area is `MultyPolygon`:

![image](https://github.com/maplibre/maplibre-gl-js/assets/1304610/d6139ec1-bf78-4cb8-a099-19477f935e46)

On the other hand, the Map view says that both `["geometry-type"]` and `$type"` are `Polygon`:

![image](https://github.com/maplibre/maplibre-gl-js/assets/1304610/605d98d6-aa8c-4f69-9b4f-989876f6da7a)

**Comment by zstadler** [comment]
### Summary of the approach taken in PR https://github.com/maplibre/maplibre-style-spec/pull/519

The style spec for [`["geometry-type"]`](https://maplibre.org/maplibre-style-spec/expressions/#geometry-type) says it _"Gets the feature's geometry type: `Point`, `MultiPoint`, `LineString`, `MultiLineString`, `Polygon`, `MultiPolygon`."_ However the implementation for tile-based geometries was identical to that of `$type` and the possible values were `Point`, `LineString`, `Polygon`. This also applied to GeoJSON sources in Maplibre-GL-JS because they are converted and stored internally as tile-based geometries.

The PR adds aligns the implementation of `["geometry-type"]` with its spec and adds the distinction between `Point` and `MultiPoint`, between `LineString` and `MultiLineString`, and between `Polygon` and `MultiPolygon`.

**Comment by louwers** [comment]
@HarelM The behavior of MapLibre Native is the same. Only `Point`, `LineString` and `Polygon` are returned. A `MultiLineString` is considered a `LineString` et cetera. It would be easy to change though, but it would be a breaking change.

**Comment by HarelM** [comment]
@louwers in both cases native returns only non multi types? or does it behave differently for `geometry-type` and `$type`?

**Comment by HarelM** [comment]
I don't know if solving a bug that aligns the product to the spec is a major or minor version change, in general, we just released version 4 of maplibre, so if this is considered a breaking change, we'll need to wait with it to version 5...
I'll bring it up in the monthly meeting tonight and see what people think.

**Comment by louwers** [comment]
It is the same for both.

```cpp
enum class FeatureType : uint8_t {
    Unknown = 0,
    Point = 1,
    LineString = 2,
    Polygon = 3
};

struct ToFeatureType {
    FeatureType operator()(const EmptyGeometry &) const { return FeatureType::Unknown; }
    template <class T>
    FeatureType operator()(const Point<T> &) const {
        return FeatureType::Point;
    }
    template <class T>
    FeatureType operator()(const MultiPoint<T> &) const {
        return FeatureType::Point;
    }
    template <class T>
    FeatureType operator()(const LineString<T> &) const {
        return FeatureType::LineString;
    }
    template <class T>
    FeatureType operator()(const MultiLineString<T> &) const {
        return FeatureType::LineString;
    }
    template <class T>
    FeatureType operator()(const Polygon<T> &) const {
        return FeatureType::Polygon;
    }
    template <class T>
    FeatureType operator()(const MultiPolygon<T> &) const {
        return FeatureType::Polygon;
    }
    template <class T>
    FeatureType operator()(const mapbox::geometry::geometry_collection<T> &) const {
        return FeatureType::Unknown;
    }
};
```

**Comment by HarelM** [comment]
Monthly meeting conversation summary:
Since this is a breaking change in terms of how maps might look there are two options to proceed:
1. Introduce this in a breaking change version (both in native and web)
2. Keep the current behavior, change the spec accordingly to align with current implementation and propose a new spec function to facilitate for this requirement (multi type filtering).

There is a strong bias towards not breaking the way styles will look, and changing the `geometry-type` implementation will cause this.
Having said that, it was agreed that `$type` is probably what is used in most styles and changing `geometry-type` to align it to the spec is OK.

I have created the following poll to solicitate more feedback:
https://github.com/maplibre/maplibre-style-spec/discussions/536

Feel free to participate there.

**Comment by wipfli** [comment]
@HarelM I lost a bit track of this conversation. Maybe you can help me with an example:

Say I have this filter:

```json
[
  "in",
  [
    "geometry-type"
  ],
  [
    "literal",
    [
        "Polygon",
        "MultiPolygon"
    ]
  ]
]
 ```
What is the current behavior, what would be the proposed new behavior?

**Comment by zstadler** [comment]
For this filter, there will be no change in behavior if the "breaking" change would be implemented.

Currently, `["geometry-type"]` never returns a `"MultiPolygon"` value, even when the geometry is a true MultiPolygon - one with multiple outer rings.
